src/backend/metal/src/command.rs

```diff
@@ -1,477 +1,634 @@
-#![allow(missing_docs)]
+use {Surface, Backend};
+use ::native;
+use ::conversions::*;
+
+use std::ops::{Deref, DerefMut, Range};
+use std::sync::{Arc};
+use std::cell::UnsafeCell;
+
+use core::{self, memory, target, pso};
+use core::{VertexCount, VertexOffset, InstanceCount, IndexCount, Viewport};
+use core::{RawSubmission};
+use core::buffer::{IndexBufferView};
+use core::image::{ImageLayout};
+use core::command::{ClearColor, ClearValue, BufferImageCopy, BufferCopy};
+use core::command::{RenderPassInlineEncoder, ImageCopy, SubpassContents};
+use core::command::{ImageResolve};
+use core::pool::{CommandPoolCreateFlags};
 
-//use cocoa::foundation::NSRange;
+use metal::*;
+use cocoa::foundation::NSUInteger;
+use block::{ConcreteBlock};
 
-use core::{self, pso, shade, state, target, texture, command};
-use core::{IndexType, VertexCount};
-use core::{MAX_VERTEX_ATTRIBUTES, MAX_CONSTANT_BUFFERS, MAX_RESOURCE_VIEWS,
-           MAX_SAMPLERS, MAX_COLOR_TARGETS};
-use core::shade::Stage;
-use {Backend, Resources};
+pub struct QueueFamily {
+}
 
-use encoder::MetalEncoder;
-use MTL_MAX_BUFFER_BINDINGS;
-use native::{Buffer, Texture, Pipeline, Rtv, Srv, Dsv};
+pub struct CommandQueue(Arc<QueueInner>);
 
-use metal::*;
+struct QueueInner {
+    queue: MTLCommandQueue,
+}
 
-use std::ptr;
-use std::collections::hash_map::{HashMap, Entry};
+unsafe impl Send for QueueInner {
+}
+
+impl Drop for QueueInner {
+    fn drop(&mut self) {
+        unsafe {
+            self.queue.release();
+        }
+    }
+}
+
+pub struct CommandPool {
+    queue: Arc<QueueInner>,
+    managed: Vec<CommandBuffer>,
+}
+
+unsafe impl Send for CommandPool {
+}
+unsafe impl Sync for CommandPool {
+}
 
 #[derive(Clone)]
-pub struct SubmitInfo {
-    pub(crate) command_buffer: MTLCommandBuffer
+pub struct CommandBuffer(Arc<UnsafeCell<CommandBufferInner>>);
```

Comment:
So the CommandBuffer can be cloned. The Metal CommandBuffer needs some ancillary data so it's not sufficient to simply add a reference to MTLCommandBuffer.

Reply:
`Clone` is only required/valid for finished command buffers to allow multi submission (probably not support atm in metal) or for convenience. The command buffer doesn't need to be in a valid state afterwards and calling commands on a cloned CommandBuffer is undefined behavior. I assume a normal clone would be enough here.

Note: I should have added clone inside `RawCommandBuffer` instead of the standard `clone`.

Reply:
I tried removing the Arc and then realized we still need it for `CommandPool.reset` to work (since MTLCommandBuffers can't actually be reset) :/